File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int len = nums.length;
        List<Integer> ans = new ArrayList<>();
        ans.add(nums[0]);

        for(int i = 1; i < len; i++){
            if(nums[i] > ans.get(ans.size() - 1))
                ans.add(nums[i]);
            
            else{
                int low = Collections.binarySearch(ans, nums[i]);

                if(low < 0)
                    low = -(low + 1);
                
                ans.set(low, nums[i]);
            }
        }

        return ans.size();
        
    }
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        ans = []

        ans.append(nums[0])

        for i in range(1, n):
            if nums[i] > ans[-1]:
                ans.append(nums[i])
            else:
                low = 0
                high = len(ans) - 1
                while low < high:
                    mid = low + (high - low) // 2
                    if ans[mid] < nums[i]:
                        low = mid + 1
                    else:
                        high = mid

                ans[low] = nums[i]

        return len(ans)

